R/bundleFiles.R

```diff
@@ -239,9 +117,75 @@ isKnitrCacheDir <- function(subdir, contents) {
 #'
 #' @export
 listBundleFiles <- function(appDir) {
-  maxDirectoryList(appDir, 0, 0, 0)
+  paths <- recursiveBundleFiles(appDir)
+
+  files <- length(paths)
+  size <- sum(file.info(file.path(appDir, paths))$size)
+
+  enforceBundleLimits(appDir, size, files)
+  list(
+    totalSize = size,
+    totalFiles = files,
+    contents = paths
+  )
 }
 
+bundleFiles <- function(appDir) {
+  listBundleFiles(appDir)$contents
+}
+
+recursiveBundleFiles <- function(dir, depth = 0) {
+  # generate a list of files at this level
+  contents <- list.files(dir, all.files = TRUE, no.. = TRUE, include.dirs = TRUE)
+
+  # exclude some well-known files/directories at root level
+  if (depth == 0) {
+    contents <- contents[!grepl(glob2rx("*.Rproj"), contents)]
+    contents <- setdiff(
+      contents,
+      c("manifest.json", "rsconnect", "packrat", "app_cache", ".Rproj.user")
+    )
+  }
+
+  # exclude renv files, knitr cache dirs, and another well-known files
+  contents <- setdiff(contents, c("renv", "renv.lock"))
+  contents <- contents[!isKnitrCacheDir(contents)]
+  contents <- setdiff(
+    contents,
+    c(".DS_Store", ".git", ".gitignore", ".quarto", ".Rhistory", ".svn")
+  )
+
+  # remove any files lines listed .rscignore
+  if (".rscignore" %in% contents) {
+    ignoreContents <- readLines(file.path(dir, ".rscignore"))
+    contents <- setdiff(contents, c(ignoreContents, ".rscignore"))
+  }
+
+  # Info for each file lets us know to recurse (directories) or aggregate (files).
+  is_dir <- dirExists(file.path(dir, contents))
+  names(is_dir) <- contents
+
+  children <- character()
+  for (name in contents) {
+    if (isTRUE(is_dir[[name]])) {
+      dirList <- recursiveBundleFiles(file.path(dir, name), depth + 1)
```

Comment:
This change removes of `getOption("rsconnect.max.bundle.size")` and `getOption("rsconnect.max.bundle.files")` as the walk progresses. This is an important check because folks occasionally use an `appDir` which is massive (a consequence of deploying `$HOME/document.Rmd`, for example).

Some historical issues with this code:
* https://github.com/rstudio/rsconnect/pull/463 - incorrect size calculations
* https://github.com/rstudio/rsconnect/issues/514 - permissions problems when enumerating
* https://github.com/rstudio/rsconnect/pull/546 - enforcement on `appFiles`

I'm not finding the original issue, but folks observed "hangs" (very long runtime) if there's no feedback until after the entire directory is processed.

Reply:
Oh bummer 😬 I'll think about how to restore this.